src/libcore/mem/mod.rs

```diff
@@ -68,9 +70,40 @@ pub use crate::intrinsics::transmute;
 /// mem::forget(file);
 /// ```
 ///
-/// The practical use cases for `forget` are rather specialized and mainly come
-/// up in unsafe or FFI code. However, [`ManuallyDrop`] is usually preferred
-/// for such cases, e.g.:
+/// This is useful when the ownership of the underlying resource was previously
+/// transferred to code outside of Rust, for example by transmitting the raw
+/// file descriptor to C code.
+///
+/// # Relationship with `ManuallyDrop`
+///
+/// While `mem::forget` can also be used to transfer *memory* ownership, doing so is error-prone.
+/// [`ManuallyDrop`] should be used instead. Consider, for example, this code:
+///
+/// ```
+/// use std::mem;
+///
+/// let mut v = vec![65, 122];
+/// // Build a `String` using the contents of `v`
+/// let s = unsafe { String::from_raw_parts(v.as_mut_ptr(), 2, v.capacity()) };
+/// // leak `v` because its memory is now managed by `s`
+/// mem::forget(v);  // ERROR - v is invalid and must not be passed to a function
+/// assert_eq!(s, "Az");
+/// // `s` is implicitly dropped and its memory deallocated.
+/// ```
+///
+/// There are two issues with the above example:
+///
+/// * If more code were added between the construction of `String` and the invocation of
+///   `mem::forget()`, a panic within it would cause a double free because the same memory
```

Comment:
Does double free in this case cause UBs even without unsafe block?

Reply:
You can't double-free without an `unsafe` block.

Reply:
In case of this UB, the wrong `unsafe` would be the call to `from_raw_parts`.

Reply:
I mean no unsafe block on the user side.

Reply:
"user side"? There's an unsafe block here, to call `from_raw_parts`. That's where you promise that it is okay to deallocate this memory. If you don't guard against panics, then you used that `unsafe` incorrectly.

Reply:
Thanks. I missed that.

Reply:
Resolving this conversation, as the text seems correct. If there is something to fix, please reopen.